Ignore residues with non-matching altlocs when only one is present.

JoaoRodrigues opened the pull request:
Addresses issue #90 .

JoaoRodrigues commented:
Congratulations @joaomcteixeira - I upgraded you to Maintainer so I can add you to reviews :)

joaomcteixeira commented:
Efficiently solves #90.

**However**, some comments:

* I added a new file to your branch to facilitate testing. See `tests/data/altloc.pdb`. *I think the file can remain in the PR if you agree*.

**Examples:**

```
pdb_selaltloc -A altloc.pdb
```

:+1: works as expected


```
pdb_selaltloc -B altloc.pdb
```

Keeps only `B` meaning that residues labeled **only** with `A` are removed. Surely this is intended by the one-tool-one-job philosophy of pdb-tools. Yet, I think we should explain this in the docstring.

```
pdb_selaltloc -C altloc.pdb
```

Because there are no `C` labels, labels `A` and `B` are removed, resulting in residues being removed entirely, silently. We need to explain this in the docstring **or** block the execution of `selaltloc` if a non-existing `altloc` is selected. However, the latter would come at a much increased computational cost for the size of the tool. I prefer the first choice.

```
pdb_selaltloc altloc.pdb
```

Removes the `altloc` labels but **does not** select the one with more occupancy, see `SER/PRO` 22. In the `tests/data/dummy.pdb` the atom `CA` belongs to the same residue, so the highest occupancy is selected. But in this case, `SER` and `PRO` have the same residue number but are different residue identities, the highest occupancy is not selected, and both residues are kept. Is this truly intended or is it an unexpected behaviour? It makes a good combo with `pdb_delinsertion` as described in #90. Just to clarify.

What are your thoughts?

**other note**

Currently, we have `dummy.pdb` that serves several tests. This creates a log of cohesion because if we modify the `dummy.pdb` to enhance the testing of some tool, several tests for other tools will break. I think we should start creating/separating tests `PDB`s one per tool. That is why I added `altloc.pdb`.

Before merging this PR, let's add as many tests as possible for this case so the exact expected behavior gets fully documented.

JoaoRodrigues commented:
@joaomcteixeira check the new commits. It should fix all the issues we discussed here (and offline):
- Picks highest occupancy if no option is provided.
- If user provides option and option does not exist for a residue, does not delete any atoms for that residue.
- Works for weird PDB files that abuse the altloc field to have different residues (e.g. 3u7t).

I added a new test and split the code in the script in two functions. Have a look please!

joaomcteixeira commented on pdbtools/pdb_selaltloc.py:
```diff
@@ -106,66 +108,111 @@ def check_input(args):
+    """
+
+    # We have to iterate multiple times
+    atom_prop = collections.defaultdict(list)
```
Despite using `collections` is more verbose and pythonic, because pdb-tools will very likely use in pipelines, and to maintain one of the main ideas of the project, use built-ins before stdlib, I suggest reverting to dict and `.setdefault` since the simpler approach is 3x faster than `defaultdict`.

```python
%%timeit
a = collections.defaultdict(list)
a['b'].append(1)
```
```
383 ns ± 10.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```


```python
%%timeit
b = {}
c = b.setdefault('b', [])
c.append(1)
```

```
127 ns ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

joaomcteixeira commented:
@JoaoRodrigues 

Very nice implementation!

I took the freedom to commit directly to your branch. Added minor updates on the main script just for micro performance improvement. Nice on the `operator.itemgetter[1]` it is actually faster than `lambda x: x[1]`.

Added the test case we discussed and extra tests covering option `-C`.

Because I was the last to commit, is on you to review and merge :wink:

JoaoRodrigues commented:
Argh. Fixed in #99
